guides/effectsize_exemplar_within.Rmd

```diff
@@ -31,93 +31,266 @@ format_num <- function(nums, sigdigits = 3) gsub("\\.$", "", formatC(nums, sigdi
   select(-`Pr(>F)`) # no need to show p-values
   
-anovaResults %>% 
-  tidy() %>% 
-  rename(F = statistic)
+anovaResults %>% as.tibble()
+```
+
+Looking at the `F` and `ges` (generalized eta squared) columns, there appear to be strong main effects for `layout` and `size` and an interaction between `layout` and `size`. However `color` and the other 2-way and 3-way interactions show only negligeable effects.
+
+
+
+### Bootstrapping
+
+<mark> Draft. Needs work. </mark>
+
+But that only gives us one value, whereas we want a confidence interval.
+
+We'll use a technique called bootstrapping, which checks the effect size for many random permutations of the data.
+
+1. Randomly sample with replacement (meaning the same value might be sampled more than once) to build a new dataset
+1. Perform the analysis on this new dataset
+1. Do that many times
+1. Sort the results for each effect and find the 95% confidence interval
+
+
+#### Each iteration
+
+Each iteration of the bootstrap samples the original dataset and runs the analysis on this permutation.
+
+```{r within-bootstrap-by-subject}
+
+# run one iteration of the bootstrap procedure
+analyzeIteration <- function(x) {
+  subjects <- unique(dataAggregated$subject)
+  
+  # select subjects at random with replacement
+  bootSubjects <- sample(subjects, length(subjects), replace=TRUE)
+  
+  # get all the results for one subject
+  # and give them a new uniwue subject id
+  getOneSubjectsData <- function(i) {
+    dataAggregated %>% 
+      filter(subject == bootSubjects[i]) %>% 
+      mutate(subject = paste(bootSubjects[i], i))
+  }
+  
+  # get all of the boostrapped subjects' data
+  bootData <- lapply(1:length(bootSubjects), getOneSubjectsData) %>% 
+    bind_rows()
+  
+  # compute the effect sizes the same way we did without bootstrapping
+  afex::aov_ez(
+    data = bootData, 
+    id = 'subject', # subject id column
+    dv = 'response_time', # dependent variable
+    within = c('layout', 'size', 'color'), # within-subject independent variables
+    between = NULL ,# between-subject independent variables
+    #fun_aggregate = mean,
+    anova_table = list(es = 'ges') # effect size = generalized eta squared
+  )$anova_table %>% 
+    as.tibble() %>% 
+    rownames_to_column('effect') %>% # put effect names in a column
+    return()
+}
+```
+
+#### Iterate
+
+The bootstrap needs to run many times to determine how a subset of the data impacts
+
+```{r within-bootstrap-bootstrap iterator}
+
+# run many iterations of the bootstrap procedure
+analyzeManyIterations = function (bootstrapIterationCount) {
+  # each core needs to reload the libraries
+  library(tidyverse)
+  library(afex)
+  lapply(1:bootstrapIterationCount, function(x) analyzeIteration(x)) %>% 
+    bind_rows()
+}
+
+# data used for bootstrapping will collapse by each subject-condition combination
+dataAggregated <- data %>% 
+  group_by(layout, size, color, subject) %>% 
+  summarize(response_time = mean(response_time))
+
```

Comment:
Would it make sense to move the definition of dataAggregated to before the line 175? This will allow readers to read from top-to-bottom.